packages/central-server/src/apiV2/userAccounts/assertUserAccountPermissions.js

```diff
@@ -19,40 +20,90 @@ export const assertUserAccountPermissions = async (accessPolicy, models, userAcc
   }
 
   const entityPermissions = await models.userEntityPermission.find({ user_id: userAccount.id });
+  const permissions = await models.permissionGroup.findManyById(
+    entityPermissions.map(ep => ep.permission_group_id),
+  );
+  const permissionsById = keyBy(permissions, 'id');
   const entities = await models.entity.findManyById(entityPermissions.map(ep => ep.entity_id));
+  const entitiesById = keyBy(entities, 'id');
   const countryCodes = entities.map(e => e.country_code).filter(c => c !== 'DL');
   if (countryCodes.length === 0) {
     // User only has access to Demo Land, and it got filtered out
     return true;
   }
+
+  // Must have Admin Panel access to all countries the user has access to
   if (!accessPolicy.allowsAll(countryCodes, TUPAIA_ADMIN_PANEL_PERMISSION_GROUP)) {
     throw new Error('Need Admin Panel access to all countries this user has access to');
   }
+  // Must have equal or higher permissions than the user has for each permission they have
+  entityPermissions.forEach(({ entity_id: entityId, permission_group_id: permissionGroupId }) => {
+    const countryCode = entitiesById[entityId].country_code;
+    const permissionGroup = permissionsById[permissionGroupId].name;
+    if (!accessPolicy.allows(countryCode, permissionGroup)) {
+      throw new Error(`Need ${permissionGroup} access to ${countryCode}`);
+    }
+  });
+
   return true;
 };
 
+const buildUserAccountRawSqlFilter = accessPolicy => {
+  const accessibleCountryCodes = [
+    ...accessPolicy.getEntitiesAllowed(TUPAIA_ADMIN_PANEL_PERMISSION_GROUP),
+    ...accessPolicy.getEntitiesAllowed(LESMIS_ADMIN_PERMISSION_GROUP),
+  ];
+  const permissionsByCountryCode = accessibleCountryCodes.reduce(
+    (obj, countryCode) => ({
+      ...obj,
+      [countryCode]: accessPolicy.getPermissionGroups([countryCode]),
+    }),
+    {},
+  );
+
+  /**
+   * Here we're building up an inner query to work out which user permissions exist that we don't
+   * have access to. Once we know that, we just filter to find the users whose ids are not in that list
+   */
+  const sql = `
+  user_account.id not in (
+    -- Inner query to detect which user_ids we don't have permission for
+    select uep.user_id from user_entity_permission uep 
+    join entity e on uep.entity_id = e.id 
+    join permission_group pg on uep.permission_group_id = pg.id 
+
+    where uep.user_id is not null and 
+    (
+      -- Either we don't have access to a country they have access to
+      e.code not in ${SqlQuery.record(accessibleCountryCodes)}
+
+      -- Or we have access to it, but not with the level of permissions they do
+${accessibleCountryCodes
+  .map(countryCode => {
+    return `      or (e.code = ? and pg.name not in ${SqlQuery.record(
+      permissionsByCountryCode[countryCode],
+    )})`;
+  })
+  .join('\n')}
+    )
+  )`;
```

Comment:
This is a pretty wild query :cowboy_hat_face: Hopefully the comments make enough sense but let me know if you feel it warrants more explanation

Reply:
Certainly taking me a bit to parse it 😅, but some minor suggestions for readability:

- ``countryCode => (`foo`)`` instead of ``countryCode => {return `foo`}``
- capitalise the SQL keywords?
- Apply De Morgan’s to the return value of the arrow function. i.e. `OR e.code <> ? OR pg.name IN ${…}

> [!WARNING]
> Suggestion below was written blindly, and its syntax hasn’t been validated!

```suggestion
  const sql = `
user_account.id NOT IN (
  -- Inner query to detect which user_ids we don’t have permission for
  SELECT uep.user_id
  FROM   user_entity_permission uep
         JOIN entity e ON uep.entity_id = e.id
         JOIN permission_group pg ON uep.permission_group_id = pg.id 
  WHERE  uep.user_id IS NOT NULL
         AND (-- Either we don't have access to a country they have access to
              e.code NOT IN ${SqlQuery.record(accessibleCountryCodes)}
              -- Or we have access to it, but not with the level of permissions they do
              ${accessibleCountryCodes.map(countryCode => (
                `OR e.code <> ?
                 OR pg.name IN ${SqlQuery.record(permissionsByCountryCode[countryCode],)`
              );}
             ).join('\n')}
)`;
```

Reply:
Oh yeah it looks much with capitalised SQL keywords!

I might push back on the rewording of `OR e.code <> 'TO' OR pg.name IN ('Public')` though. To me, it reads more intuitively to say `entity is one we have access to AND permission group is one we do not have access to`. Thoughts?

I didn't know about 'De Morgan's' rule though, thanks for sharing that.

Reply:
Yeah, actually, when you put it next to how we’d more naturally say it English, I agree that refactor is less intuitive